src/Swarm/Language/Syntax.hs

```diff
@@ -147,6 +146,37 @@ east = V2 1 0
 west :: V2 Int64
 west = V2 (-1) 0
 
+-- | The direction for moving vertically down = @V2 0 0@.
+down :: V2 Int64
+down = V2 0 0
+
+-- | The 'applyTurn' function gives the meaning of each 'Direction' by
+--   turning relative to the given vector or by turning to an absolute
+--   direction vector.
+applyTurn :: Direction -> V2 Int64 -> V2 Int64
+applyTurn = dirApplyTurn . dirInfo
+
+-- | Mapping from heading to their corresponding cardinal directions
+--   only directions with a 'dirAbs` value are mapped
+cardinalDirs :: M.Map (V2 Int64) Direction
+cardinalDirs =
+  M.fromList
+    . map (\d -> (fromJust . dirAbs . dirInfo $ d, d))
+    . filter (isJust . dirAbs . dirInfo)
```

Comment:
You can make this shorter and safer (needs `TupleSection` above and change of imports to `(fromMaybe, mapMaybe)`)
```suggestion
    . mapMaybe (\d -> (,d) <$> dirAbs (dirInfo d))
```